### PR TITLE
OpenNMT example

### PR DESCRIPTION
**Pull request by adamlerer**
The following OpenNMT features are implemented:

- multi-layer bidirectional RNNs with attention and dropout
- data preprocessing
- saving and loading from checkpoints
- inference (translation) with batching and beam search

Not yet implemented:

- word features
- multi-GPU
- residual connections

### CODE REVIEW

File: OpenNMT/onmt/Optim.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import math
+import torch.optim as optim
+
+class Optim(object):
+
+    def _makeOptimizer(self):
+        if self.method == 'sgd':
+            self.optimizer = optim.SGD(self.params, lr=self.lr)
+        elif self.method == 'adagrad':
+            self.optimizer = optim.adagrad(self.params, lr=self.lr)
```
Comment (MaximumEntropy) [inline OpenNMT/onmt/Optim.py]: Not sure if this has been fixed on another PR branch but there are typos in the optimization methods besides SGD ( they aren't capitalized)

Comment (adamlerer) [inline OpenNMT/onmt/Optim.py]: Thanks, fixed.

File: OpenNMT/onmt/modules/GlobalAttention.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        """
+        input: batch x dim
+        context: batch x sourceL x dim
+        """
+        targetT = self.linear_in(input).unsqueeze(2)  # batch x dim x 1
+
+        # Get attention
+        attn = torch.bmm(context, targetT).squeeze(2)  # batch x sourceL
+        if self.mask is not None:
+            attn.data.masked_fill_(self.mask, -math.inf)
```
Comment (szagoruyko) [inline OpenNMT/onmt/modules/GlobalAttention.py]: there's no math.inf in py2

**Comment by szagoruyko** [comment]
really nice work @adamlerer ! en-fr model on benchmark-1M gets to 12 perplexity after 1 epoch and generates decent predictions. I think we can merge this, just 2 things:
 * math.inf needs a fix in py2 as I pointed above
 * would be nice to have some pretrained models available. I will upload mine after it converges.

**Comment by szagoruyko** [comment]
uploaded en-fr model state_dict in hdf5 to AWS (4.85 ppl): https://s3.amazonaws.com/pytorch/h5models/model_enfr_e13_4.85.hkl (320MB)
One thing, checkpointing should be done via `.state_dict()` and `.load_state_dict()` for model and optimizer.

**Comment by bmccann** [comment]
@szagoruyko, thanks for uploading that model! I had a few questions about the training. Why do you emphasize using checkpointing with .state_dict() and .load_state_dict()? Did you make any other changes to train that model? Have you been able to get BLEU scores?

File: OpenNMT/train.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    model.train()
+    return total_loss / total_words
+
+
+def trainModel(model, trainData, validData, dataset, optim):
+    print(model)
+    model.train()
+    for p in model.parameters():
+        p.data.uniform_(-opt.param_init, opt.param_init)
+
```
Comment (bmccann) [inline OpenNMT/train.py]: I think the two lines above will reinitialize parameters even if the model is loaded from a checkpoint. Potential fix at https://github.com/adamlerer/examples/pull/2

**Comment by apaszke** [comment]
Can we pack the data and upload it to our s3 instead of checking it in?

**Comment by adamlerer** [comment]
Yes, that's the plan. I'm going to upload these models to our so and add
then to the Readme.

**Comment by szagoruyko** [comment]
@bmccann no modifications to the code, haven't tried to compute BLEU scores yet. state_dict is the preferred way of checkpointing, it allows you to make changes to model definition and keep the parameters. In case of optim, it also prevents from the state breaking when core PyTorch changes.
@adamlerer I think @apaszke means the text data?

**Comment by quilby** [comment]
I apologize if this isnt the place to post this, but has anyone tried training this model on a large dataset? I tried training it on the europarl v7 en->de dataset. The preprocessing works fine but when I start training I get:


> Namespace(batch_size=64, brnn=False, brnn_merge='concat', cuda=True, curriculum=False, >data='data/ep7-s-train.pt', dropout=0.3, epochs=13, input_feed=1, layers=2, learning_rate=1, > >learning_rate_decay=0.5, log_interval=50, max_generator_batches=32, max_grad_norm=5, >optim='sgd', param_init=0.1, pre_word_vecs_dec=None, pre_word_vecs_enc=None, rnn_size=500, >save_model='model', start_decay_at=8, start_epoch=1, train_from=None, word_vec_size=500)
>Loading data from 'data/ep7-s-train.pt'
> * vocabulary size. source = 50004; target = 50002
> * number of training sentences. 20
> * maximum batch size. 64
> Building model...
> * number of parameters: 84820002
NMTModel (
  (encoder): Encoder (
    (word_lut): Embedding(50004, 500, padding_idx=0)
    (rnn): LSTM(500, 500, num_layers=2, dropout=0.3)
  )
  (decoder): Decoder (
    (word_lut): Embedding(50002, 500, padding_idx=0)
    (rnn): StackedLSTM (
      (dropout): Dropout (p = 0.3)
      (layer_0): LSTMCell(1000, 500)
      (layer_1): LSTMCell(500, 500)
    )
    (attn): GlobalAttention (
      (linear_in): Linear (500 -> 500)
      (sm): Softmax ()
      (linear_out): Linear (1000 -> 500)
      (tanh): Tanh ()
    )
    (dropout): Dropout (p = 0.3)
  )
  (generator): Sequential (
    (0): Linear (500 -> 50002)
    (1): LogSoftmax ()
  )
)

>Traceback (most recent call last):
  File "train.py", line 280, in <module>
    main()
  File "train.py", line 276, in main
    trainModel(model, trainData, validData, dataset, optim)
  File "train.py", line 204, in trainModel
    train_loss = trainEpoch(epoch)
  File "train.py", line 163, in trainEpoch
    batchOrder = torch.randperm(len(trainData))
RuntimeError: must be strictly positive at /data/users/soumith/miniconda2/conda-bld/pytorch-cuda80-0.1.8_1486041592240/work/torch/lib/TH/generic/THTensorMath.c:1473

**Pull request merged**